### PR TITLE
Upgrade to 1.x standards

### PR DESCRIPTION
**Pull request by marcomc**
#### WORK IN PROGRESS ####
bring up to speed with harness-base-php v1.x

- store `nginx` and `lighthouse` docker image definitions and basic configurations in the `_base` harness, but keep the disables
- `nginx` and `lighthouse` are finetuned and enabled in the specific harness definitions that will use them (for now only node-spa)

### CODE REVIEW

File: src/_base/application/overlay/auth.json.twig [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+            "password": "{{ repo.password }}"
+        }{% if not loop.last %},{% endif %}
+      {% endfor %}
+    },
+    "github-oauth": {
+      {% if @('composer.auth.github') %}
+        "github.com": "{{ @('composer.auth.github') }}"
+      {% endif %}
+    }
+}
```
Comment (andytson-inviqa) [inline src/_base/application/overlay/auth.json.twig]: we don't use composer in node, so this auth.json isn't needed, along with the composer.* variables related to it if in other files

File: src/_base/application/skeleton/README.md.twig [hunk 2]
```diff
@@ -17,18 +17,23 @@ Please follow the below steps to get started, if you encounter any issues instal
...
 3. Run `ws install`
 
-Once installed, the site should be available at [https://{{ @('hostname') }}](https://{{ @('hostname') }}).
+{% include blocks ~ 'websites.md.twig' %}
 
+**Additional services**:
+
+{% if @('services.jenkins.enabled') -%}
+- Jenkins: [https://{{ @('jenkins.external_host') }}](https://{{ @('jenkins.external_host') }})
+{% endif %}
```
Comment (andytson-inviqa) [inline src/_base/application/skeleton/README.md.twig]: do we have jenkins service here? not needed for node projects

File: src/_base/application/overlay/auth.json.twig (hunk 1 above)
Comment (marcomc) [reply src/_base/application/overlay/auth.json.twig]: fixed: removed references and functions related to Composer. thank you

File: src/_base/harness/config/pipeline.yml [hunk 3]
```diff
@@ -1,47 +1,74 @@
 command('app build'):
   env:
-    COMPOSE_PROJECT_NAME: = @('namespace')
+    HAS_JENKINS_RUNNER: "= @('services.jenkins-runner.enabled') ? 'true' : 'false'"
```
Comment (andytson-inviqa) [inline src/_base/harness/config/pipeline.yml]: spryker-specific, not node

File: src/_base/application/skeleton/README.md.twig (hunk 2 above)
Comment (marcomc) [reply src/_base/application/skeleton/README.md.twig]: I thought we were using Jenkins to run Lighthouse for the node-spa harness based projects...

File: src/_base/harness/config/pipeline.yml (hunk 3 above)
Comment (marcomc) [reply src/_base/harness/config/pipeline.yml]: fixed

File: src/_base/application/skeleton/README.md.twig (hunk 2 above)
Comment (andytson-inviqa) [reply src/_base/application/skeleton/README.md.twig]: Currently it's just `ws lighthouse` being run from the CI of the project's Jenkinsfile. I don't think we'd make people suffer it's bloat per project for lighthouse.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
+# Harness Base PHP for [Workspace]
```
Comment (kierenevans) [inline README.md]:
```suggestion
# Harness Base Node for [Workspace]
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+A [Workspace] harness is a way to ship files to a project without being part of the project.
+
+In this repository are a set of harnesses that have been created for the PHP language, reducing the maintenance overhead
+of the individual harnesses greatly. This is due to the "base" PHP harness being used as a common set of templates.
```
Comment (kierenevans) [inline README.md]:
```suggestion
In this repository are a set of harnesses that have been created for the NodeJS language, reducing the maintenance overhead
of the individual harnesses greatly. This is due to the "base" NodeJS harness being used as a common set of templates.
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+  ```
+7. Adjust each downstream repository's default branch so people can easily follow the README
+
+[Workspace]: https://github.com/my127/workspace
+[GitHub Changelog Generator]: https://github.com/github-changelog-generator/github-changelog-generator
+[inviqa/harness-node]: https://github.com/inviqa/harness-node
+[inviqa/harness-node-spa]: https://github.com/inviqa/harness-node-spa
+[inviqa/harness-viper]: https://github.com/inviqa/harness-viper
+[my127/my127.io]: https://github.com/my127/my127.io
+[version specific upgrade instructions]: https://github.com/inviqa/harness-base-php/blob/1.1.x/UPGRADE.md
```
Comment (kierenevans) [inline README.md]:
```suggestion
[version specific upgrade instructions]: https://github.com/inviqa/harness-base-node/blob/0.2.x/UPGRADE.md
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+12. Ask for someone else to test the pull request.
+13. Once the pull request has been merged to the default branch of the project, remind the project team to apply the changes with the `ws harness update existing` or `ws harness update fresh` commands.
+
+## Helm charts
+
+Each harness deploys:
+* A "console" pod for running one-off commands
+* A NGINX/PHP-FPM "webapp" pod for running the php-based application and serving web requests
+* A service to route to the "webapp" pods
+* An ingress definition to route via the "webapp" service
```
Comment (kierenevans) [inline README.md:73]: Needs updating to reflect what we do here

File: README.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+## Deployment
+
+### Deploy Script
+
+The [deploy script](./deploy) does a similar thing but the end result is output to `src/<framework>`, where it is then
+committed as a publish commit to a temporary build branch.
+
+A "subtree-split" is then performed which outputs a directory for each folder into a "publish" folder, where it is then
+force pushed to the individual harness repositories' `1.1.x` branch.
```
Comment (kierenevans) [inline README.md:132]: 1.1.x and references to 1.1.1/1.2.0 will need to be updated based on the next release version number

File: UPGRADE.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
+# Upgrade Guide
```
Comment (kierenevans) [inline UPGRADE.md:1]: This file will need updating based on version numbers and actual breaking changes (if any)

File: src/_base/README.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
+# PHP Harness for [Workspace]
+
+To use this harness:
+
+1. Install [Workspace]
+2. Run `ws create <projectName> inviqa/php:v1.1.1`
```
Comment (kierenevans) [inline src/_base/README.md]:
```suggestion
# NodeJS Harness for [Workspace]

To use this harness:

1. Install [Workspace]
2. Run `ws create <projectName> inviqa/node:v0.2.0`
```

File: src/_base/application/skeleton/README.md.twig [hunk 11]
```diff
@@ -39,59 +39,83 @@ To remove the development environment, run `ws destroy`.
...
+The frontend build should be automatically done as part of bringing up the environment.
 
-For development mode it's expected that `yarn dev` starts up a development
-node server that serves static assets
+To trigger a rebuild, run `ws frontend build`.
 
 For pipeline and production modes, it's expected that `yarn build` will build
 distribution artifacts in the ./build directory. This will then be copied into
-an nginx image to serve the assets staticly.
+an node image to serve the assets staticly.
```
Comment (kierenevans) [inline src/_base/application/skeleton/README.md.twig:48]: nginx in the case of a single page application, node in the case of a node app serving both BE and FE from the same place?

File: src/_base/docker-compose.yml.twig [hunk 12]
```diff
@@ -1,20 +1,36 @@
...
+{% endif %}
 {% endfor %}
 networks:
   private:
     external: false
   shared:
-    external:
-      name: my127ws
-{% if @('docker-compose.sync.strategy') == 'mutagen' %}
+    name: my127ws
```
Comment (kierenevans) [inline src/_base/docker-compose.yml.twig]:
```suggestion
    external:
      name: my127ws
```
Still needed due to linking up with https://github.com/my127/workspace/blob/8f424b9a50142328a034728cad23ad5cb64bf34b/home/bin/ws.service#L27

File: src/_base/harness/attributes/common.yml [hunk 13]
```diff
@@ -1,53 +1,234 @@
...
-    image_pull_config: = docker_config(@('docker.registry'))
     additional_schema_locations: https://inviqa.github.io/kubernetes-json-schema/schema
-
     feature:
-      sealed_secrets: true
+      # Note: be very careful considering disabling this, as in most cases
+      # it causes the secrets in it to be stored plaintext on filesystem
+      # or in helm chart repositories
+      # requires sealed-secrets k8s operator
+      sealed_secrets: false
```
Comment (kierenevans) [inline src/_base/harness/attributes/common.yml]:
```suggestion
      sealed_secrets: true
```
We meant to turn this on by default in php harness v1.0.0, but forgot.

File: src/_base/harness/attributes/common.yml [hunk 14]
```diff
@@ -1,53 +1,234 @@
 attributes.default:
+  app:
+    # build - static|dynamic
+    #   dynamic - volumes are mounted and the build step is run once the containers have started
+    #   static  - app is copied into console, built, then the resulting build is copied into the web image
```
Comment (kierenevans) [inline src/_base/harness/attributes/common.yml]: Description may need updating

File: src/_base/harness/attributes/common.yml [hunk 15]
```diff
@@ -1,53 +1,234 @@
...
+      enabled: true
+    compose:
+      file_version: '3.7'
+      host_volume_options: "= ':' ~ (@('delegated-volumes') ? 'delegated' : 'cached')"
+    # If using gitops helm chart repositories, it's recommended to put this configuration
+    # in the values.yml in there now instead of project application repositories
+    image_pull_config: = @('docker.config')
+    image:
+      node: = 'node:' ~ @('node.version') ~ '-slim'
+      nginx: nginx:1.17-alpine
```
Comment (kierenevans) [inline src/_base/harness/attributes/common.yml]:
```suggestion
      nginx: nginx:1.21-alpine
```

based on supported tags on https://hub.docker.com/_/nginx

File: src/_base/application/overlay/Jenkinsfile.twig [hunk 16]
```diff
@@ -1,43 +1,89 @@
...
     triggers { cron(env.BRANCH_NAME == '{{ @('git.main_branch') }}' ? 'H H(0-6) * * *' : '') }
     stages {
+{% if @('jenkins.tests.isolated') %}
+        stage('Setup') {
+            steps {
+                sh 'ws harness download'
+                sh 'ws harness prepare'
+                milestone(10)
+            }
+        }
```
Comment (kierenevans) [inline src/_base/application/overlay/Jenkinsfile.twig:18]: Missing something to build and bring up a container - maybe `node`?

File: src/_base/harness/config/functions.yml [hunk 17]
```diff
@@ -167,3 +175,47 @@ function('slugify', [text]): |
...
+      ',',
+      array_map(
+        function ($pool) use ($hostname) {
+          return 'tcp://' . $hostname .':' . $pool['port'] . '/status';
+        },
+        $pools
+      )
+    );
+  = $text;
+
```
Comment (kierenevans) [inline src/_base/harness/config/functions.yml]:
```suggestion
```
Not needed if no php-fpm

File: src/_base/harness/scripts/rebuild.sh [hunk 18]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+run docker compose down --rmi local --volumes --remove-orphans --timeout 12
+
+passthru ws cleanup built-images
+
+run rm -f .my127ws/.flag-built .my127ws/.flag-console-built
```
Comment (kierenevans) [inline src/_base/harness/scripts/rebuild.sh]:
```suggestion
run rm -f .my127ws/.flag-built
```

File: src/node-spa/.ci/sample-dynamic/workspace.yml [hunk 19]
```diff
@@ -1,10 +1,16 @@
-workspace('ci-node-spa-sample-dynamic'):
+workspace('ci-node-sample-dynamic'):
```
Comment (kierenevans) [inline src/node-spa/.ci/sample-dynamic/workspace.yml]:
```suggestion
workspace('ci-node-spa-sample-dynamic'):
```

File: src/_base/harness/config/commands.yml [hunk 20]
```diff
@@ -1,37 +1,62 @@
...
+    HAS_ASSETS:           = @('aws.bucket') !== null and @('aws.bucket') !== '' ? 'yes':'no'
     COMPOSE_PROJECT_NAME: = @('namespace')
+    COMPOSE_DOCKER_CLI_BUILD: = @('docker.buildkit.enabled') ? '1':'0'
+    DOCKER_BUILDKIT:          = @('docker.buildkit.enabled') ? '1':'0'
   exec: |
     #!bash(workspace:/)
+    set -- all
+    source .my127ws/harness/scripts/enable.sh
+
+command('enable console'):
```
Comment (kierenevans) [inline src/_base/harness/config/commands.yml]: Do we need this one, or should we adapt it to bring up a node container?

File: src/viper/harness/config/confd.yml [hunk 21]
```diff
@@ -2,20 +2,24 @@
 confd('harness:/'):
   - { src: docker/image/node/Dockerfile }
   - { src: docker/image/node/root/etc/supervisor/conf.d/client.conf }
-  - { src: docker/image/node/root/lib/task/build.sh }
   - { src: docker/image/node/root/lib/task/build/gateway.sh }
   - { src: docker/image/node/root/lib/task/build/client.sh }
```
Comment (kierenevans) [inline src/viper/harness/config/confd.yml:8]:
```suggestion
  - { src: docker/image/node/root/lib/task/build/client.sh }
  - { src: docker/image/lighthouse/Dockerfile }
  - { src: docker/image/lighthouse/root/app/run.sh }
```
Do we need any of these?

**Review by kierenevans — COMMENTED** [review]
Great work, thank you 👍 

I've taken a look through this and saw a few things, hopefully easily solvable

With src/*/docker/image/node/root/lib/task/start/dev.sh.twig deleted, the `command` used in src/_base/_twig/docker-compose.yml/service/node.yml.twig and src/viper will need updating. What do we need to start by default?

`grep -R image_pull_config src/` might need some cleaning up - app vs helm.

`src/viper/_twig/docker-compose.yml/service/node.yml.twig` references `docker-compose.sync.native.options` still

File: src/viper/harness/config/confd.yml (hunk 21 above)
Comment (marcomc) [reply src/viper/harness/config/confd.yml:8]: I didn't want to add the lighthouse stuff until I was able to run the test Viper harness alone

File: src/_base/harness/config/commands.yml (hunk 20 above)
Comment (marcomc) [reply src/_base/harness/config/commands.yml]: in some of the conversation, I think @andytson-inviqa vented the idea that in future we might introduce the `console` container into node harnesses, so I didn't remove all `console` related commands yet, until making sure that we don't need them. also, I was keeping them as a reference of what kind of operations we need to run  inside a base node container that is the one that has to work as a kind of `console`

File: src/_base/application/overlay/Jenkinsfile.twig (hunk 16 above)
Comment (marcomc) [reply src/_base/application/overlay/Jenkinsfile.twig:18]: didn't get to debug Jenkins build yet... I want to make sure I can run `./test <harness> static` locally first, to have a comparison to know what to expect to happen in Jenkins

**Comment by andytson-inviqa** [comment]
will do this another way

**Pull request closed without merging**